src/QMCHamiltonians/CoulombPBCAB.cpp

```diff
@@ -181,7 +181,10 @@ CoulombPBCAB::Return_t CoulombPBCAB::evaluateWithIonDerivs(ParticleSet& P,
                                                            ParticleSet::ParticlePos_t& pulay_terms)
 {
   forces = 0.0;
-  Value  = evalLRwithForces(P) + evalSRwithForces(P) + myConst;
+  if (ComputeForces)
```

Comment:
evaluteWithIonDerivs exists to force computation of ion derivatives without the if(ComputeForces) flag.  This function should always call evalXwithForces.

Reply:
Let me write it down, so that we get mutual understanding of the logic:

The "forces=yes" attribute indicates, whether we want to include force contributions from a particular hamiltonian term, such as IonIon, ElecElec, ElecIon, or PseudoPot. PseudoPot is considered in another class, and we do not concern ElecElec or ElecIon (bare) forces in this case. The main concern here is IonIon, which may contain AA and AB terms. So, adding "forces=yes/no" to IonIon should allocate an Ewald handler with/without gradients, respectively.

The ACForce estimator seems to be a separate player, calling evaluateIonDerivatives() irregardless of the forces-attribute. That means it is up to particular hamiltonians to control whether they are on/off, returning energy/energy+forces, respectively. (For convenience, I think every physically motivated hamiltonian should be on by default.)

The problem arises, when evaluateWithIonDerivatives() is requesting a gradient from a "forces=no"-hamiltonian, because none have been allocated. Possible fixed include

use logic to make handling of Ewald gradients foolproof inside hamiltonian classes (eg CoulombPBCAA/CoulombPBCAB)
always allocate Ewald gradients, when ACForce is present
The current proposition falls into category 1. I don't see a problem with it in IonIon (AB) case, but I agree there could be a problem with ElecIon hamiltonians. It is only logical that evalXwithForces be called every time, but then there should be extra care either in allocating the gradients or requesting them. Either that or we confirm that the gradients are unnecessary, if the the hamiltonian is "excluded" from the forces with "forces=no". Am I making sense?

To me this does not sound like an overly complicated problem, but I thought I'd ask your opinion before making the next move.

Reply:
Okay, actually the problem is rather with PseudoPot hamiltonian than IonIon, but anyway there is something that needs fixing